apps/consulting/posts/improvements-to-bsr-matrix-multiplication-in-pytorch.md

```diff
@@ -0,0 +1,297 @@
+    name="fig-4"
+    alt="Two plots displaying sparsity ratio (as a percentage) on the horizontal axis, and speedup
+    over the baseline on the vertical.  Three lines are plotted on both upper and lower plots
+    indicating the block sizes used 16, 32, and 64."
+/>
+
+Comparison to the [dense baseline][#fig-5] is also very promising. Block sizes of 32 and 64 are
+faster than dense for almost all sparsity ratios, and block size 16 crossing the 1x boundary at 70%
+sparsity. The `float32` performance relative to the dense baseline for larger block sizes is
+starting to approach ideal behavior. That is, when the sparse kernel does not have any overhead
+associated with metadata lookups we could expect it to compute the result with a speedup
+proportional to the inverse of the sparsity; at 80% sparsity only 1/5 of the tensor values are
+specified and it should run about 5x faster. We are very close to this mark with ~3.5x and ~4.5x,
+for block size 32 and 64 respectively.
+
+<img src="posts/improvements-to-bsr-matrix-multiplication-in-pytorch/bsr_dense_mm_float32_vs_dense.png"
+    name="fig-5"
+    alt="A plot displaying sparsity ratio (as a percentage) on the horizontal axis, and speedup over
+    the dense baseline on the vertical.  There is a horizontal line marking the speedup ratio of 1.0
+    above which the sparse subject takes less time to execute than the dense counterpart.  Three lines are
+    plotted indicating block sizes used for the sparse tensor of 16, 32, and 64."
+/>
+
+Now we look at the performance for half precision data types. The new kernel provides between 20-50x
+speedup over the [older sparse implementation][#fig-6]. While a speedup of this size is always a
+notable result, we have known that the implementation supporting half precision data types would not
+be optimal. However, as cuSPARSE did not support these data types a working implementation was
+created by composing calls into the dense kernel for each sparse block. This approach suffers from a
+significant penalty due to launching many CUDA kernels.
+
+<img src="posts/improvements-to-bsr-matrix-multiplication-in-pytorch/bsr_dense_mm_half_vs_sparse.png"
+    name="fig-6"
+    alt="A plot displaying sparsity ratio (as a percentage) on the horizontal axis, and speedup over
+    the older sparse implementation on the vertical.  Three lines are plotted indicating block sizes
+    used for the sparse tensor of 16, 32, and 64."
+/>
+
+Comparing performance for half precision to the [dense baseline][#fig-7] we see a clear win.
+Using a block size of 64, the triton implementation begins to outperform the dense matrix multiply
+at 60% sparsity. Unfortunately, this does not hold for smaller block sizes with 32 reach >1x
+speedup above 80% sparsity, and block size 16 only passing this point above 90%.
+
+<img src="posts/improvements-to-bsr-matrix-multiplication-in-pytorch/bsr_dense_mm_half_vs_dense.png"
```

Comment:
Some general feedback about graphs.
1. Data points are generally preferable to jagged line graphs. This makes it clear where the measured data points are. It may also help with overlapping lines that currently make the data a bit hard to read.
2. Some of the speedup graphs are hard to read near 1x as the scale is so large. It may be better to have the largest sparsities go off the scale, or maybe even a log-y plot.
3. Since so much importance is given to the 1x speedup point, it would be good to annotate it on the graph for each line.